daml-lf/spec/daml-lf-1.rst

```diff
@@ -4023,7 +4261,37 @@ Error functions
 
 * ``ERROR : ∀ (α : ⋆) . 'Text' → α``
 
-  Throws an error with the string as message.
+  Throws an error with the string as message. See the evaluation rule
+  ``EvExpError`` for precise semantics.
+
+* ``USER_EXCEPTION : 'Text' → 'UserException'``
+
+  This is not exposed in the expression language. It is only here so we can use
+  it in the value language.
+
+* ``USER_EXCEPTION_MESSAGE : 'UserException' → 'Text'``
+
+* ``ARITHMETIC_EXCEPTION : 'Text' → 'ArithmeticException'``
+
+  This is not exposed in the expression language. It is only here so we can use
+  it in the value language.
+
+* ``ARITHMETIC_EXCEPTION_MESSAGE : 'ArithmeticException' → 'Text'``
+
+* ``PRECONDITION_EXCEPTION : ∀ (τ : ⋆) . τ → 'PreconditionException'``
```

Comment:
I think we should really avoid putting an any/polymorphic type into an exception. Partly because of serialization concerns, partly because we don't allow for existential types in general... I'd be happier with this having a `Text` payload (that tells you which template failed), and if someone wants a better exception they can throw it themselves (using fancy ensure statements).

The other thing is I think this should be called `ENSURE_EXCEPTION` or something, to match the surface syntax.